pkg/infra/metrics/service.go

```diff
@@ -55,5 +59,62 @@ func (im *InternalMetricsService) Run(ctx context.Context) error {
 	return ctx.Err()
 }
 
-func ProvideRegisterer() prometheus.Registerer        { return prometheus.DefaultRegisterer }
-func ProvideRegistererForTest() prometheus.Registerer { return prometheus.NewRegistry() }
+func ProvideRegisterer(cfg *setting.Cfg) prometheus.Registerer {
+	if cfg.IsFeatureToggleEnabled(featuremgmt.FlagGrafanaAPIServer) {
+		return legacyregistry.Registerer()
+	}
+	return prometheus.DefaultRegisterer
+}
+
+func ProvideGatherer(cfg *setting.Cfg) prometheus.Gatherer {
+	if cfg.IsFeatureToggleEnabled(featuremgmt.FlagGrafanaAPIServer) {
+		return newK8sGathererWrapper(legacyregistry.DefaultGatherer)
+	}
+	return prometheus.DefaultGatherer
+}
+
+func ProvideRegistererForTest() prometheus.Registerer {
+	return prometheus.NewRegistry()
+}
+
+func ProvideGathererForTest(reg prometheus.Registerer) prometheus.Gatherer {
+	// the registerer provided by ProvideRegistererForTest
+	// is a *prometheus.Registry, so it also implements prometheus.Gatherer
+	return reg.(*prometheus.Registry)
+}
+
+type k8sGathererWrapper struct {
```

Comment:
since this is not k8s specific (just the motivation)... maybe different name?  we can also change that when the feature flag is actually removed